lib/Sema/CSSimplify.cpp

```diff
@@ -1410,6 +1410,35 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     case ConstraintKind::ArgumentConversion:
     case ConstraintKind::OperatorArgumentTupleConversion:
     case ConstraintKind::OperatorArgumentConversion:
+      if (type1->is<InOutType>() && !type2->is<InOutType>() && typeVar2) {
+        // Try to upgrade type2 to an InOutType
+        auto inOutType1 = type1->getAs<InOutType>();
+        auto typeVar3 = createTypeVariable(getConstraintLocator(locator),
+                                           typeVar2->getImpl().getOptions());
+        auto inOutType3 = InOutType::get(typeVar3);
+        
+        TypeMatchOptions subFlags = flags;
+        subFlags -= TMF_GenerateConstraints;
+        // First, see if we can actually replace typeVar2 with an InOutType.
```

Comment:
OK, that was the part I wasn't quite sure about in my approach, looking at the implementation of `matchTypes` I couldn't see a reason for the CS to get changed if you remove `TMF_GenerateConstraints`, but you know the part far better than me ;-). I'll have a look at adding a new flag to type variable then